### PR TITLE
Add pcap export functionnality

### PR DESCRIPTION
**Pull request by jeanthom**
This PR adds pcap export functionnality to lcsniff. Some of the code added here could be reused to implement extcap, as suggested in https://github.com/lambdaconcept/usbsniffer/issues/1

### CODE REVIEW

**Comment by povauboin** [comment]
Hello,

It crashes for me when loading a previously saved file and exporting to pcap.

File: src/pcapexport.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		auto packet = record.packet(i);
+		if (packet) {
+			writePacket(*packet);
+		}
+	}
+}
+
+void PCapExport::writePacket(USBPacket& packet)
+{
+	auto packetData = packet.recordData().second;
```
Comment (povauboin) [inline src/pcapexport.cpp]: crashes on this line

File: src/usbgroup.cpp [hunk 2]
```diff
@@ -69,3 +69,13 @@ bool USBGroup::matchForFilter(const USBProxyFilter *filter) const
...
 }
+
+int USBGroup::packetCount() const
+{
+    return m_count;
+}
+
+USBPacket* USBGroup::packet(int i) const
+{
+    return &m_first[i];
```
Comment (povauboin) [inline src/usbgroup.cpp]: this looks suspicious to me

File: src/usbrecord.h [hunk 3]
```diff
@@ -17,6 +17,8 @@
 #define RECORD_LENGTH   5
 #define RECORD_SUMMARY  6
 
+class USBPacket;
```
Comment (povauboin) [inline src/usbrecord.h]: also, I dont really like the fact that USBRecord depends on USBPacket, considering that USBPacket already inherits from USBRecord

**Review by povauboin — COMMENTED** [review]
perhaps a better way to handle this would be to go through the USBItem tree, and look for USBPacket elements.This way we can directly use USBPacket.m_Packet for writing a pcap chunk, and no longer need these functions:

```
USBPacket* USBRecord::packet(int i) const
{
    return nullptr;
}
```

**Comment by jeanthom** [comment]
Hi @povauboin,

Thanks for your precious feedback. I tried another approach which exposes packets from the USBAggregator class. This way we avoid creating dependency between USBRecord and USBPacket in the wrong way, we don't need to add methods to USBRecord subclasses and we can correctly access all the packets from a USBGroup.

File: src/usbgroup.cpp (hunk 2 above)
Comment (jeanthom) [reply src/usbgroup.cpp]: I was thinking that USBPackets were in adjascent memory locations but it turned out they weren't... Code removed in https://github.com/lambdaconcept/usb2sniffer-qt/pull/27/commits/89523ec40e9990e5afb8ca6b0c154420675f9c7f.

File: src/usbrecord.h (hunk 3 above)
Comment (jeanthom) [reply src/usbrecord.h]: Dependency removed in https://github.com/lambdaconcept/usb2sniffer-qt/pull/27/commits/89523ec40e9990e5afb8ca6b0c154420675f9c7f.

File: src/pcapexport.cpp (hunk 1 above)
Comment (jeanthom) [reply src/pcapexport.cpp]: Probably related to the bug introduced in `src/usbgroup.cpp`, the bug doesn't seem to come from this function but from the caller giving the wrong pointer.

File: src/pcapexport.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	for (int i = 0; i < aggregator.count(); i++) {
+		writePacket(*aggregator.value(i));
+	}
+}
+
+void PCapExport::writeGlobalHeader(void)
+{
+	pcap_hdr_t header;
+
+    header.magic_number = 0xA1B2C3D4;
```
Comment (desowin) [inline src/pcapexport.cpp]: Why not use nanosecond resolution magic (0xA1B23C4D)?

Comment (jeanthom) [reply src/pcapexport.cpp]: Wasn't aware of it, now using it.

Comment (desowin) [reply src/pcapexport.cpp]: Well, in the latest commit there is invalid magic. You use there 0xA1B2C34D while it should be 0xA1B23C4D.

Comment (jeanthom) [reply src/pcapexport.cpp]: Fixed!

**Comment by povauboin** [comment]
Thanks, this is much cleaner now.

It requires the development version of wireshark to open the pcap, I am correct ?

**Comment by jeanthom** [comment]
Yes the code for USBLL is only in Wireshark master. You can get nightly builds for Windows & macOS here : https://www.wireshark.org/download/automated/

**Pull request merged**